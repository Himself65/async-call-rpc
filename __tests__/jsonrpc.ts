import {
    Request,
    SuccessResponse,
    ErrorResponse,
    hasKey,
    isJSONRPCObject,
    isObject,
    ErrorResponseMapped,
} from '../src/utils/jsonrpc'
test('Request', () => {
    expect(Request('id', 'method', ['param1', 'param2'], 'stack')).toMatchSnapshot('req1')
    expect(Request('id2', 'method', { param1: 'abc', param2: 'def' }, 'stack')).toMatchSnapshot('req2')
})

test('SuccessResponse', () => {
    expect(SuccessResponse('id2', undefined)).toMatchSnapshot('undefined')
    expect(SuccessResponse('id3', { data: 'result' })).toMatchSnapshot('success')
})

test('ErrorResponse', () => {
    expect(ErrorResponse('id2', -123, 'message', { _data_: 1 })).toMatchSnapshot('normal error')
    expect(ErrorResponse('id2', -123.456, 'message')).toMatchSnapshot('invalid code')
    expect(ErrorResponse(undefined, 0, '')).toMatchSnapshot()
    expect(ErrorResponse('', -0, '')).toMatchSnapshot()
    expect(ErrorResponse('', NaN, '')).toMatchSnapshot()
    expect(ErrorResponse.InvalidRequest('id')).toMatchSnapshot('invalid req')
    expect(ErrorResponse.MethodNotFound('id')).toMatchSnapshot('method not found')
    expect(
        ErrorResponse.ParseError(new Error(), () => ({
            code: 2345,
            message: 'My message',
            data: { my_data: true },
        })),
    ).toMatchSnapshot('parse error')
})

test('hasKey', () => {
    expect(hasKey({ a: 1 }, 'a')).toBe(true)
    expect(hasKey({ a: 1 }, 'b')).toBe(false)
})

test('isJSONRPCObject', () => {
    expect(isJSONRPCObject(Request('id', '', [], ''))).toBe(true)
    expect(isJSONRPCObject(SuccessResponse('id', undefined))).toBe(true)
    expect(isJSONRPCObject(ErrorResponse('id', 0, 'msg', 'stack'))).toBe(true)
    expect(isJSONRPCObject({})).toBe(false)
    expect(isJSONRPCObject(undefined)).toBe(false)
    expect(isJSONRPCObject({ jsonrpc: '1.0' })).toBe(false)
    expect(isJSONRPCObject({ jsonrpc: '2.0', params: 1 })).toBe(false)
})

test('isObject', () => {
    expect(isObject('')).toBeFalsy()
    expect(isObject({})).toBeTruthy()
    expect(isObject(null)).toBeFalsy()
})
