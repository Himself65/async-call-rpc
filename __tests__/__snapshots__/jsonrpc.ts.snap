// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorResponse: internal error 1`] = `
Object {
  "error": Object {
    "code": -32603,
    "data": Object {
      "stack": "",
      "type": "Error",
    },
    "message": "Internal errormsg",
  },
  "id": "id",
  "jsonrpc": "2.0",
}
`;

exports[`ErrorResponse: invalid code 1`] = `
Object {
  "error": Object {
    "code": -124,
    "data": Object {
      "stack": "stack",
      "type": "Object",
    },
    "message": "message",
  },
  "id": "id2",
  "jsonrpc": "2.0",
}
`;

exports[`ErrorResponse: invalid error object 1`] = `
Object {
  "error": Object {
    "code": 123,
    "data": Object {
      "stack": "stack",
      "type": "Error",
    },
    "message": "",
  },
  "id": "id",
  "jsonrpc": "2.0",
}
`;

exports[`ErrorResponse: invalid params 1`] = `
Object {
  "error": Object {
    "code": -32602,
    "data": Object {
      "stack": "",
      "type": "Error",
    },
    "message": "Invalid params",
  },
  "id": "id",
  "jsonrpc": "2.0",
}
`;

exports[`ErrorResponse: invalid req 1`] = `
Object {
  "error": Object {
    "code": -32600,
    "data": Object {
      "stack": "",
      "type": "Error",
    },
    "message": "Invalid Request",
  },
  "id": "id",
  "jsonrpc": "2.0",
}
`;

exports[`ErrorResponse: method not found 1`] = `
Object {
  "error": Object {
    "code": -32601,
    "data": Object {
      "stack": "",
      "type": "Error",
    },
    "message": "Method not found",
  },
  "id": "id",
  "jsonrpc": "2.0",
}
`;

exports[`ErrorResponse: normal error 1`] = `
Object {
  "error": Object {
    "code": -123,
    "data": Object {
      "stack": "stack",
      "type": "Error",
    },
    "message": "message",
  },
  "id": "id2",
  "jsonrpc": "2.0",
}
`;

exports[`ErrorResponse: parse error 1`] = `
Object {
  "error": Object {
    "code": -32700,
    "data": Object {
      "stack": "stack",
      "type": "Error",
    },
    "message": "Parse error",
  },
  "id": null,
  "jsonrpc": "2.0",
}
`;

exports[`Request: req1 1`] = `
Object {
  "id": "id",
  "jsonrpc": "2.0",
  "method": "method",
  "params": Array [
    "param1",
    "param2",
  ],
  "remoteStack": "stack",
}
`;

exports[`Request: req2 1`] = `
Object {
  "id": "id2",
  "jsonrpc": "2.0",
  "method": "method",
  "params": Object {
    "param1": "abc",
    "param2": "def",
  },
  "remoteStack": "stack",
}
`;

exports[`SuccessResponse: success 1`] = `
Object {
  "id": "id3",
  "jsonrpc": "2.0",
  "result": Object {
    "data": "result",
  },
}
`;

exports[`SuccessResponse: undefined-dropping 1`] = `
Object {
  "id": "id2",
  "jsonrpc": "2.0",
  "result": null,
  "resultIsUndefined": true,
}
`;

exports[`SuccessResponse: undefined-keeping 1`] = `
Object {
  "id": "id",
  "jsonrpc": "2.0",
  "result": null,
}
`;
